### ISSUE TITLE
NetBSD build fails - frexpl conflicting types

### ISSUE THREAD

**Issue by bremensaki** [ksh93/ksh#528]
I've hit a strange build issue on NetBSD 9.3 (possibly other versions). On only some architectures, compilation halts with a type conflict for frexpl - on others it's fine.

```
In file included from /usr/pkgsrc/shells/ast-ksh/work/src/lib/libast/sfio/sfhdr.h:302:0,
                 from /usr/pkgsrc/shells/ast-ksh/work/src/lib/libast/port/mc.c:31:
./FEATURE/float:181:33: error: conflicting types for 'frexpl'
 extern __MANGLE__ _ast_fltmax_t frexpl __PROTO__((_ast_fltmax_t, int*));
                                 ^~~~~~
In file included from ./FEATURE/float:85:0,
                 from /usr/pkgsrc/shells/ast-ksh/work/src/lib/libast/sfio/sfhdr.h:302,
                 from /usr/pkgsrc/shells/ast-ksh/work/src/lib/libast/port/mc.c:31:
/usr/include/math.h:361:13: note: previous declaration of 'frexpl' was here
 long double frexpl(long double, int *);
             ^~~~~~
```
I've tracked this down in the build logs to a config test failing, but I don't know exactly how these tests are defined, or where. Pointers on how I can replicate this in isolation would be great.

For example - on sparc:
```
iffe: test: is -lm a library ... yes
iffe: test: is frexp a library function ... yes
iffe: test: is frexpl a library function ... no
iffe: test: is ldexp a library function ... yes
iffe: test: is ldexpl a library function ... yes
```
But on an amd64 system:
```
iffe: test: is frexpl a library function ... yes
```

I've done the hideously ugly hack of reversing the logic in the tests of _lib_frexpl in features/float and comp/frexpl.c and getting a successful build.

There's a full failed build log from an alpha system (not one of mine), where the same issue manifests here: http://ftp.ziaspace.com/pub/pkgsrc/reports/9.0_2022Q2/20220630.1248//ast-ksh-20120801nb2/build.log

The fact that it fails on both sparc and alpha is curious as it kind of eliminates it being a 32/64 bit or endianness problem. Given the existence of binaries being available in pkgsrc for amd64, i386 and aarch64 it's probably fine on those, but failing at least on alpha, earmv7hf and sparc. I've only personally replicated the success and failure on amd64 and sparc respectively.

**Comment by McDutchie** [comment]
You seem to be trying to build a 93u+ 2012-08-01 version. That is not supported here as this repo maintains a new 93u+m version with a lot of bugs fixed and a lot of ancient baggage removed (including all that `__MANGLE__` stuff, see a34e83192b62493eb0e08874c855c386c1be6615). But none of us ever had access to a SPARC or Alpha machine to test it, so it would be interesting to see current results from those. Could you try again with the [93u+m/1.0.3 release](https://github.com/ksh93/ksh/releases/tag/v1.0.3), please?

**Comment by bremensaki** [comment]
I first encountered this problem while building CDE 2.5.0 which has 1.0.3 incorporated into the tree currently - I was having no issues with building on amd64, but it failed on sparc.

I went back to the pkgsrc version to try just ksh in isolation and found the same bug in place there too. I'll try a standalone build of the current release, but I don't expect different results from the CDE build that started me down this path.

Where's this test defined in the source tree?

**Comment by McDutchie** [comment]
The test is defined here: https://github.com/ksh93/ksh/blob/536696782c3ba9d29f2db5d11b4b3f0aa8f1961b/src/lib/libast/features/float#L4

The `lib` iffe command checks for the presence of a function in the system libraries and the `npt` iffe command checks for the need to define a prototype for a function. Those are standard iffe tests and I don't know why iffe is not detecting frexpl in the sparc or alpha system libraries. Could that be a bug in those libraries? I have no way of testing as I don't have access to those architectures.

But hopefully we can at least make sure that the replacement implementation works properly. Apparently, on those architectures, `_ast_fltmax_t` is not equivalent to `long double`. But since `long double` is specified by the C90 standard which we now require, we can just change the code to use that instead.

Does the following patch (to 93u+m) fix the problem?

```diff
diff --git a/src/lib/libast/comp/frexpl.c b/src/lib/libast/comp/frexpl.c
index 2d80877f3..cc4345c3c 100644
--- a/src/lib/libast/comp/frexpl.c
+++ b/src/lib/libast/comp/frexpl.c
@@ -42,13 +42,13 @@ NoN(frexpl)

 #else

-static _ast_fltmax_t	pow2tab[LDBL_MAX_EXP + 1];
+static long double	pow2tab[LDBL_MAX_EXP + 1];

 static int
 init(void)
 {
 	register int		x;
-	_ast_fltmax_t		g;
+	long double		g;

 	g = 1;
 	for (x = 0; x < elementsof(pow2tab); x++)
@@ -69,12 +69,12 @@ init(void)

 #undef	frexpl

-extern _ast_fltmax_t
-frexpl(_ast_fltmax_t f, int* p)
+extern long double
+frexpl(long double f, int* p)
 {
 	register int		k;
 	register int		x;
-	_ast_fltmax_t		g;
+	long double		g;

 	INIT();

@@ -139,8 +139,8 @@ frexpl(_ast_fltmax_t f, int* p)

 #undef	ldexpl

-extern _ast_fltmax_t
-ldexpl(_ast_fltmax_t f, register int x)
+extern long double
+ldexpl(long double f, register int x)
 {
 	INIT();
 	if (x < 0)
diff --git a/src/lib/libast/features/float b/src/lib/libast/features/float
index 8edfc9577..b25a79eb9 100644
--- a/src/lib/libast/features/float
+++ b/src/lib/libast/features/float
@@ -878,13 +878,13 @@ tst	- note{ missing floating point limits }end output{
 		printf("extern double		frexp(double, int*);\n");
 	#endif
 	#if !_lib_frexpl || _npt_frexpl
-		printf("extern _ast_fltmax_t	frexpl(_ast_fltmax_t, int*);\n");
+		printf("extern long double	frexpl(long double, int*);\n");
 	#endif
 	#if !_lib_ldexp || _npt_ldexp
 		printf("extern double		ldexp(double, int);\n");
 	#endif
 	#if !_lib_ldexpl || _npt_ldexpl
-		printf("extern _ast_fltmax_t	ldexpl(_ast_fltmax_t, int);\n");
+		printf("extern long double	ldexpl(long double, int);\n");
 	#endif

 		return 0;
diff --git a/src/lib/libast/sfio/sfhdr.h b/src/lib/libast/sfio/sfhdr.h
index a2ef22089..a24bb265b 100644
--- a/src/lib/libast/sfio/sfhdr.h
+++ b/src/lib/libast/sfio/sfhdr.h
@@ -987,7 +987,7 @@ extern int		errno;
 #define frexpl		frexp
 #endif
 #if !_lib_frexpl
-extern Sfdouble_t	frexpl(Sfdouble_t, int*);
+extern long double	frexpl(long double, int*);
 #endif
 #endif
 #ifndef ldexpl
@@ -995,7 +995,7 @@ extern Sfdouble_t	frexpl(Sfdouble_t, int*);
 #define ldexpl		ldexp
 #endif
 #if !_lib_ldexpl
-extern Sfdouble_t	ldexpl(Sfdouble_t, int);
+extern long double	ldexpl(long double, int);
 #endif
 #endif

```

**Comment by McDutchie** [comment]
Here is an additional patch to features/float that replaces the default frexpl and ldexpl tests with tests that compile a regular program and check if they work. Can you check if this works around the problem of iffe not detecting these in the system libraries on NetBSD/SPARC?
```diff
diff --git a/src/lib/libast/features/float b/src/lib/libast/features/float
index 8edfc9577..23c071bff 100644
--- a/src/lib/libast/features/float
+++ b/src/lib/libast/features/float
@@ -1,7 +1,25 @@
 set	nooptimize
 set	stdio FEATURE/standards
 hdr	float,limits,math,values
-lib,npt	frexp,frexpl,ldexp,ldexpl,finite,finitel,isinfl,isnanl,copysign,copysignl,powf,pow,powl FEATURE/standards math.h -lm
+lib,npt	frexp,ldexp,finite,finitel,isinfl,isnanl,copysign,copysignl,powf,pow,powl FEATURE/standards math.h -lm
+lib	frexpl -lm note{ frexpl present and works }end link{
+		#include <math.h>
+		int main(void)
+		{
+			int e;
+			long double f = frexpl(123.456789,&e);
+			return !(e==7 && f>0.964506 && f<0.964507);
+		}
+	}end
+lib	ldexpl -lm note{ ldexpl present and works }end link{
+		#include <math.h>
+		int main(void)
+		{
+			long double f = ldexpl(0.96450616406250000434141611549421213567256927490234375, 7);
+			return !(f>=123.4567890 && f<123.4567891);
+		}
+	}end
+npt	frexpl,ldexpl
 lib	fpclassify -lm note{ fpclassify present and works }end link{
 		#include <sys/types.h>
 		#include <sys/stat.h>
```

**Comment by bremensaki** [comment]
These compiles take ages, especially when I reset to a clean tree every time - but I can confirm a couple of things. First, the second patch, the test changes, work properly - but interestingly make the prototype checks fail where they didn't before.
```
iffe: test: frexpl present and works ... yes
iffe: test: ldexpl present and works ... yes
iffe: test: is stdlib.h a header ... yes
iffe: test: is unistd.h a header ... yes
iffe: test: is frexpl a symbol that needs a prototype ... yes
iffe: test: is ldexpl a symbol that needs a prototype ... yes
```
Pre-patch we got this, even when frexpl wasn't detected as a library function:
```
iffe: test: is frexp a symbol that needs a prototype ... no
iffe: test: is frexpl a symbol that needs a prototype ... no
iffe: test: is ldexp a symbol that needs a prototype ... no
iffe: test: is ldexpl a symbol that needs a prototype ... no
```
The compile still ends with a type conflict, and I'm about to start a build with the first patch included, which I expect to succeed anyway. I'm going to see if I can get an emulator up on the weekend to run this a bit more efficiently than the real hardware.
```
./FEATURE/float:128:22: error: conflicting types for 'frexpl'
 extern _ast_fltmax_t frexpl(_ast_fltmax_t, int*);
                      ^~~~~~
In file included from ./FEATURE/float:32:0,
                 from /usr/local/src/ksh-1.0.3/src/lib/libast/sfio/sfhdr.h:174,
                 from /usr/local/src/ksh-1.0.3/src/lib/libast/port/mc.c:27:
/usr/include/math.h:361:13: note: previous declaration of 'frexpl' was here
 long double frexpl(long double, int *);
             ^~~~~~
In file included from /usr/local/src/ksh-1.0.3/src/lib/libast/sfio/sfhdr.h:174:0,
                 from /usr/local/src/ksh-1.0.3/src/lib/libast/port/mc.c:27:
```

I have the full build output if you think it'd be of any use, but everything else seems pretty mundane and looks to match amd64 builds.

**Comment by bremensaki** [comment]
1.0.3 builds cleanly in isolation on NetBSD/sparc with both patches in place. I haven't thoroughly tested the resulting ksh binary, but it does start and work as a basic shell.

I'm setting up an emulated environment to build things a bit faster so I can see if I can narrow things down a bit in iffe. There's something strange going on if simply moving the npt tests to another line changes their results. I'm starting to suspect there's some subtle bug lower down in the dependency chain at this point.

**Comment by McDutchie** [comment]
> There's something strange going on if simply moving the npt tests to another line changes their results.

That bit is entirely understandable. The npt tests determine if a prototype definition is needed for the functions in questions. They weren't simply moved; their outcome depends on the outcome of the lib tests, and those lib tests had a workaround applied so they now work on NetBSD/sparc using a different detection method.

The real question is, why is the generic iffe lib test not detecting frexpl()?

I've now installed a QEMU sparc emulator and installed NetBSD 9.3/sparc (32-bit) on it. I can reproduce the build failure. (FYI, it's dog slow.)

One thing I've noticed is this:
```
$ uname -a
NetBSD netsparc.inlv.org 9.3 NetBSD 9.3 (GENERIC) #0: Thu Aug  4 15:30:37 UTC 2022  mkrepro@mkrepro.NetBSD.org:/usr/src/sys/arch/sparc/compile/GENERIC sparc
$ strings -a /usr/lib/libm.so|grep frexp
frexpf
frexp
$NetBSD: s_frexpl.c,v 1.3 2013/02/12 21:40:19 martin Exp $
$NetBSD: s_frexpf.c,v 1.10 2007/08/21 20:12:27 drochner Exp $
$NetBSD: s_frexp.c,v 1.13 2008/09/28 18:54:55 christos Exp $
frexp
frexpf
$ 
```

So, frexpl is not found as a symbol name using `strings`(1), though the corresponding .c file is apparently compiled in. Whereas frexp and frexpf are found as symbol names. What gives, there? Could this point to the reason why the generic iffe lib test fails?

If the same `strings -a` is done grepping for `ldexpl`, it is found -- and also, iffe finds that using the generic test.

Also, on NetBSD/x86_64, `strings -a` does find frexpl in libm.so.

So I think the problem is with the NetBSD libm on sparc, here.

I will commit the workaround regardless, as it doesn't do any harm and it has the additional benefit of doing a basic verification that the functions work as expected.
